[FLINK-37481] Add new multi way join operator

gustavodemorais opened the pull request:
_The PR has a working version for inner joins. Outer joins and different types of input (no unique key, join on unique key) are in progress._

<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

We currently always rely on a chain of binary joins with StreamingJoinOperator when there are multiple joins in a query. This adds a multi way join operator that is able to process multiple inputs at the same time.

A multi way join operator is capable of joining multiple inputs without any intermediate state, thus, considerably reducing the total state necessary for chained join operations. This only adds the operator, PRs for the other required changes will follow.

Out of scope: we could improve the time complexity of this algorithm with an adaptive joining strategy, this can be done in a second interaction. For now, we iterate through all inputs and check one join condition across inputs for all possibilities. 

## Brief change log
  - Added multi-way join operator
  - (ongoing) Added multi join state handler
  - Added harness tests for the operator


## Verifying this change
  - Operator harness test for inner joins
  - (ongoing) harness test for outer joins

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,474 @@
+package org.apache.flink.table.runtime.operators.join.stream;
```
missed Apache license
imho as a result the reason of license check ci failure

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,474 @@
+import java.util.List;
+
+/**
+ * Streaming multi-way join operator which supports inner join and left/right/full outer join. It
+ * eliminates the intermediate state necessary for a chain of multiple binary joins. In other words,
```
From one side the code uses JoinRelType, from another side the comment says nothing about `SEMI`, `ANTI`, `ASOF` joins and probably others...
What are the plans for those within this PR and in future?

gustavodemorais replied:
Yeah, I'll refactor so we use an internal JoinType with the supported joins. For this initial implementation of the multi-way join operator, I deliberately focused on INNER and LEFT joins, as outlined in the FLIP's "Limitations and Scope" section. The goal was to first address the most common use cases we've seen causing large state issues and try to somehow limit the scope

gustavodemorais replied:
Refactored to use internal join type and remove calcite dep https://github.com/apache/flink/pull/26313/files#diff-0209feea41d82c2a944d1cb848d3a00bbdbeb4ba7c4accc189d52f441e0a02a1R57

gustavodemorais commented on flink-table/flink-table-runtime/pom.xml:
```diff
@@ -137,7 +137,13 @@ under the License.
+        <dependency>
+            <groupId>org.apache.calcite</groupId>
+            <artifactId>calcite-core</artifactId>
+            <version>1.34.0</version>
```
removed the calcite dep

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,474 @@
+    private final List<InternalTypeInfo<RowData>> inputTypes;
+    private final MultiJoinCondition multiJoinCondition;
+    private final boolean[] filterNulls;
+    private final long[] stateRetentionTime;
```
I'm confirming what we level of granularity we're able to support here according to the parser

gustavodemorais replied:
Technically, every input can have a different state ttl, e.g. 

```
SELECT /*+ STATE_TTL('posts'='6h', 'users'='2d') */ *
FROM posts
JOIN users ON posts.user_id = users.id;
```

When I get to the optimizer, I'll confirm it works for N inputs. But either way, I think it's safer and correct for the operator to support a different state ttl for each input.

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,474 @@
+    }
+
+    private void processElement(int inputId, RowData input, long timestamp) throws Exception {
+        inputId = inputId - 1; // Convert to 0-based index
```
Makes sense. The reason I have to do it at all is because the AbstractInput requires input to start from 1 when we create it.

I've simplified the logic to call [processElement with id - 1](https://github.com/apache/flink/pull/26313/files#diff-0209feea41d82c2a944d1cb848d3a00bbdbeb4ba7c4accc189d52f441e0a02a1R379) and all the logic inside the operator is 0-based. Now processElement [doesn't change the inputId](https://github.com/apache/flink/pull/26313/files#diff-0209feea41d82c2a944d1cb848d3a00bbdbeb4ba7c4accc189d52f441e0a02a1R114). It's similar to [what we do here](https://github.com/apache/flink/blob/33de4ea7739dcee4dd40f2217754bf2b17e3b589/flink-runtime/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperatorV2.java#L552).

I personally like the solution above, but other solutions are doing `id - 1` all over the operator or creating arrays with an empty first position. What do you think?

gustavodemorais commented:
I've refactored and added an example + key comments to the algorithm. Ongoing: I'll be going through possible smaller optimizations to make the algorithm more efficient. 

Since the operator requires a complex logic, now'd be a good time to take a look at the core algorithm before more optimizations are added.

davidradl commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,806 @@
+ * partitioning. It eliminates the intermediate state necessary for a chain of multiple binary
+ * joins. In other words, it reduces the total amount of state necessary for chained joins. As of
+ * time complexity, it performs better in the worst cases where the number of records in the
+ * intermediate state is large but worst than reordered binary joins when the number of records in
```
It is difficult to read this sentence as it uses worst in 2 senses. Maybe replace the last sentence with :

```
In other words, it reduces the total amount of state necessary for chained joins. As of
time complexity, it performs better in cases where the number of records in the intermediate state is large but less well with reordered binary joins and the number of records in the intermediate state is small.
```

gustavodemorais replied:
I'll rewrite it 👍

snuyanzin commented on flink-table/flink-table-runtime/pom.xml:
```diff
@@ -137,7 +137,7 @@ under the License.
 			<scope>test</scope>
 		</dependency>
-    </dependencies>
+	</dependencies>
```
can we remove it from change log?

gustavodemorais replied:
Aren't we squashing all commits? It'll then not be in the commit/changelog, since I removed the change

Or do you think it'd be better to merge a a couple of separate commits and not squash? I could then rebuild the commits when we're done with the reviews

dawidwys commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/keyselector/AttributeBasedJoinKeyExtractor.java:
```diff
@@ -0,0 +1,299 @@
+        // Iterate through join attributes for the current depth.
+        // Key (leftAttrRef) points to previous input (< depth).
+        // Value (rightAttrRef) points to current input (== depth).
+        for (Map.Entry<AttributeRef, AttributeRef> entry : attributeMapping.entrySet()) {
```
Why do we need to iterate over all the entries? Can't we precompute indices for a given join depth?
I'd imagine in a pseudocode:
```
val mask = joinMasks.get(depth);
RowData value = currentRows[depth];
RowData key = mask.apply(value);
```

I am a bit afraid of doing all the expensive computations for every record.

gustavodemorais replied:
That's a valid concern. It's not so straightforward as you have in your pseudocode because we build the index/key based on all current rows < depth. Basically, we build the key based on the left side of the join before iterating through the rows of depth, "the right side".

Mostly, for join conditions with unique key, we'll extract the key only for one record. However, I agree we had unnecessary expensive recomputations here in case we do not have primary keys. I refactored the code so we precompute these extractors only once and use them for all records 🙂 See: https://github.com/apache/flink/pull/26313/commits/0aa71e40ebd4120bf17d3a19e4acbb39217626f6

dawidwys commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,809 @@
+
+/**
+ * Streaming multi-way join operator which supports inner join and left outer join, right joins are
+ * transformed into right joins by the optimizer. It only supports a combination of joins that joins
```
Is the sentence correct?

> right joins are transformed into right joins by the optimizer

dawidwys replied:
what is the value of the sentence then? right joins into right joins? Or do you mean `sequence of right join two input joins`?

gustavodemorais replied:
Updated

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/MultiJoinStateViews.java:
```diff
@@ -0,0 +1,356 @@
+
+        @Override
+        public Iterable<RowData> getRecords(RowData mapKey) throws Exception {
+            Map<RowData, RowData> uniqueKeyToRecordMap = recordState.get(mapKey);
```
> mapKey != uniqueKey
Exactly. Maybe I join rename mapKey to joinKey?

> Imo, it should be possible to add a test case which produces wrong results if mapKey is not equal to the uniqueKey
Yeah, actually yes. It's a bit tricky with the setup, but it's a good test. I'll refactor the setup and look into that

gustavodemorais replied:
I think I'll rename mapKey to joinKey. I think that makes it clearer, what the key is

gustavodemorais replied:
done

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/MultiJoinStateViews.java:
```diff
@@ -0,0 +1,356 @@
+     *
+     * <p>Stores data as {@code MapState<MapKey, Record>}.
+     */
+    private static final class JoinKeyContainsUniqueKey implements MultiJoinStateView {
```
> The state is scoped by the mapKey already 

I don't think that's always the case. If we partition the input using only one key and the join condition contains two keys. For example.

`SELECT * FROM T1 JOIN T2 on T1.k1 = T2.k2 JOIN T3 on T2.k2 = T3.k3 AND T2.k99 == T3.k99`

Let's suppose (T2.k2, T2.k99) and (T3.k3, T3.k99) are composite unique keys from T2 and T3. We will partition rows across multi-join operators based on T1.k1, T2.k2 and T2.k3. 

>There will be at most 1. record per key

Inside the operator, we can use the composite unique key as the key of the map and store one record per key.

What do you think? Either way, I'll add tests for multiple conditions like this.

dawidwys replied:
I think your explanation makes sense. Let's add the tests though.

gustavodemorais commented:
I currently have these tests: JoinHasUniqueKeys join InputHasUniqueKey join InputHasUniqueKey

Working on these tests
- Non-equi join
- Multi keys
- InputHasNoUniqueKey
- 4 way join:  JoinHasUniqueKeys join InputHasUniqueKey join inputHasNoUniqueKey join JoinHasUniqueKeys

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,810 @@
+            return false;
+        }
+
+        // --- Calculate mapKey for reading state at this depth using left side ---
```
Done

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,810 @@
+            // This ensures the count reflects matches found in *subsequent* levels (depth+1
+            // onwards).
+            if (isLeftJoin) {
+                associations[depth] = 0;
```
Ops, true!

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,810 @@
+            int inputId,
+            RowData[] currentRows,
+            int[] associations,
+            JoinPhase phase)
```
Yes, we do this, but in several loops. The logic is the same, but it divides the algorithm between "searching" for the input while calculating the amount of associations and continuing the join with the input inside current rows.

```

CALCULATE_MATCHES -> Recuse through inputs until we reach inputId level, which triggers EMIT_RESULTS... We continue recursing to the right of inputId level and eventually the recursion returns with all the combinations to the right with input id have been checked ->
CALCULATE_MATCHES again for the next combinations of rows to the left of inputId -> EMIT_RESULTS with the new combination to the left + input... eventually returns  ->
CALCULATE_MATCHES ... and so on. 
```

My idea with the JoinPhase was to make that easier to understand, since we repeat the same thing for different paths/combinations of rows. But it clearly wasn't the case :D I also had a version where this variable was basically a `boolean shouldEmit`. We could also call it `boolean afterInputId`, `boolean processedInput`, `boolean withInput` or smth like that.

That said, I'll look into it and see if I can come up with other ideas and will share here.

gustavodemorais replied:
After trying out some stuff, I went forward with the boolean approach to try to make it easier to understand. Take a look https://github.com/apache/flink/pull/26313/commits/6b3d0ab301fc73774c88662c46c734ec65ea9df8

gustavodemorais commented:
Refactored the comments and tried to focus more on the reason behind each key step. Let me know what you think @dawidwys. Specifically, let me know if there are more places you think we can improve it or tone it down

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,830 @@
+        // If the current depth is the one where the triggering input record arrived,
+        // now process the input record itself with the current combination of rows we are at.
+        if (isInputLevel(depth, inputId)) {
+            processInputRecord(depth, input, inputId, currentRows, associations, matched);
```
Am I correct that we call this method `processInputRecord` only here for condition
```java
if (isInputLevel(depth, inputId))
```

then the question if we need to pass both `depth` and `inputId`?

Asking since based on body of `isInputLevel`
```java
    private boolean isInputLevel(int depth, int inputId) {
        return depth == inputId;
    }
```
we can drop one of them from arg list

snuyanzin replied:
Also a question: I wonder if we need `isInputLevel` method at all?
Asking since I notice that there is a number of places where we have same conditions in `if` without usage of this method

gustavodemorais replied:
> Asking since I notice that there is a number of places where we have same conditions in if without usage of this method

Oh yeah, true. I've refactored, so the other two places use it. Let me know if I forgot one

> we can drop one of them from arg list
What could we drop? I didn't get this part

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,830 @@
+            // This information is crucial for determining if null padding is needed later.
+            if (isLeftJoin) {
+                updateAssociationCount(
+                        depth, associations, shouldIncrementAssociation(phase, input));
```
it looks like we do not need to calculate `shouldIncrementAssociation(phase, input)` on each loop iteration and can do it before the loop or did I miss anything?

gustavodemorais replied:
We increment the number of associations for ´current depth - 1` for each record, in case they match. We first check if the join conditions matches for each record inside the loop, if it doesn't, we don't want to increment the number of associations and just continue to the next record.

```
if (matchesCondition(depth, currentRows)) {
                matched = true;
            } else {
                continue;
            }
```

SteveStevenpoor commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/MultiJoinStateViews.java:
```diff
@@ -0,0 +1,359 @@
+    private static final class InputSideHasNoUniqueKey implements MultiJoinStateView {
+
+        // stores map in the mapping <MapKey, Map<Record, Count>>
+        private final MapState<RowData, Map<RowData, Integer>> recordState;
```
If we use Rocks or Forst state backend, we will put a record in rocksdb with a key which is composite [flink-key + join_key] and the value will be the map of {record, counter}. In case we have large keys with the same join key we will need to get very large value. Even more, on each addRecord operation you need to get a very big value and deserialize a very big value, update and put it back. It seems to me that with bigger state we eventually will start to struggle from those get operations. I suggest to reuse JoinRecordStateViews from the StreamingJoinOperator or adapt their logic of using join key as part of a flink key. It will help to reduce value size and improve state efficiency

gustavodemorais replied:
Hey Steve, thanks for the comment! It's a valid concern. However, we can't use the join key as the operator key/partition key due to some limitations, see https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=350784112#FLIP516:MultiWayJoinOperator-CommonPartitioningKeyRequired. Basically, if we use all joins keys, the input data for different inputs would not also land in the right operator and would be partitioned to different operators. We can only use a subset of the common attribute used across all join conditions. That's why I had to write a new state view.

That said, your comment make me wonder if I could reduce the amount read from state by creating a composite key. In this case, something like ("operator-key__join-key" -> integer). I'll think a bit about it.

SteveStevenpoor replied:
Hey! Thanks for the clarification. From what I understand, the restriction means that we don't need to use all joins keys. We just need to use the common key and everything will work like a charm. Also your earlier example in review comments:

> SELECT * FROM T1 JOIN T2 on T1.k1 = T2.k2 JOIN T3 on T2.k2 = T3.k3 AND T2.k99 == T3.k99

Do I understand right that this example is also violate restrictions since there's no single key that all inputs can be partitioned on? Please correct me if I’m missing something.

If my point is correct and clear for you, I think it could be worth reusing JoinRecordStateView to reduce value size.

And also I still be happy to help or split the work if it's useful. As far as I know the multi-join operator is not obvious in terms of both algorithmic side and architecture.

gustavodemorais replied:
> Do I understand right that this example is also violate restrictions since there's no single key that all inputs can be partitioned on? Please correct me if I’m missing something.

This example doesn't violate it because we have one common key by transitivity across all three tables. T1.k1 for T1, T2.K2 for T2 and T3.K3 for T3. We basically need one value that can be used for patitioning for all inputs. This would violate: `SELECT * FROM T1 JOIN T2 on T1.k1 = T2.k2 JOIN T3 on T2.k10 = T3.k20 AND T2.k99 == T3.k99`

> And also I still be happy to help or split the work if it's useful. 

That's great to hear. Glad to get some help to tackle these. Would you be open to contribute to the optimizer or mostly to the runtime? We have to adjust the optimizer rules and think that'd be a easiest way of splitting the load, see [FLINK-37814 ](https://issues.apache.org/jira/browse/FLINK-37814).

SteveStevenpoor replied:
> Yes, everything does work with only one common key. However, we have multiple map levels to try to reduce the amount of records to reprocess. My first version actually only used the common key, and I've added this second level as the results of some discussions with @dawidwys on how to optimize the operator.

You do reduce the amount of records but the size of these records will be very big. On each get we will have to take a whole map of keys with common join key, deserialize it, update, serialize and put back which is quite costy. If we have 1 000 000 keys with the same join key, we can't be sure that the map you get from state on addRecord will even fit into memory.

> This example doesn't violate it because we have one common key by transitivity across all three tables. `T1.k1 for T1, T2.K2 for T2 and T3.K3 for T3`. We basically need one value that can be used for patitioning for all inputs. Let me know if that makes sense to you. This would violate it: `SELECT * FROM T1 JOIN T2 on T1.k1 = T2.k2 JOIN T3 on T2.k10 = T3.k20 AND T2.k99 == T3.k99`

Thank you! I understand now. But I still don't get why we don't reuse JoinRecordStateViews though. I don't see any drawbacks or restrictions to do this. Please let me know if I missing something.

> That's great to hear. Glad to get some help to tackle these. Would you be open to contribute to the optimizer or mostly to the runtime? We have to adjust the optimizer rules and think that'd be a easiest way of splitting the load, see [FLINK-37814 ](https://issues.apache.org/jira/browse/FLINK-37814).

Thank you! I can contribute anything related to multi-join. Can you please assign me to [FLINK-37814 ](https://issues.apache.org/jira/browse/FLINK-37814)?

SteveStevenpoor replied:
Also I'd like to share couple of thoughts about this example again:

> SELECT * FROM T1 JOIN T2 on T1.k1 = T2.k2 JOIN T3 on T2.k2 = T3.k3 AND T2.k99 == T3.k99

Say parallelism is 2. k1 = 1, k2 = 1, T2.k99 = 2. Let 1 be partitioned into first subtask. Currently, Flink will partition T2 by composite key (k2, k99). Let (1,2) be partitioned in the second subtask in our case. So 1 from T1 will never be joined with (1,2) from T2 in this situation. From what I understand, you suggest to partition T2 by only k2, overriding Calcite's logical distribution. And it will resolve the issue, I totally agree with it. However, I advise caution about going against Calcite's traits since it can lead to several unnecessary complications. Given this, maybe we should consider this example as "common partition key" violation, at least on the first implementation iteration? Let me know your thoughts on it, please.

gustavodemorais replied:
> Thank you! I understand now. But I still don't get why we don't reuse JoinRecordStateViews though. I don't see any drawbacks or restrictions to do this. Please let me know if I missing something.

InputSideHasNoUniqueKey in JoinRecordStateViews can simply use MapState<Row, Count> since the operator key is already the join key. Our Multi-Join Operator has the common key as the operator key, and not necessarily the join key, so we need to have MapState<JoinKey, Map<Row, Count>>. We need this extra level to be able to do [state.getRecords(joinKey)](https://github.com/apache/flink/blob/f0e6cd145d89dd4dde32e3724e57aa0167fecd26/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java#L564.)and filter down the records to reprocess. How would we be able to do that if we simply use MapState<Row, Count>?

> Given this, maybe we should consider this example as "common partition key" violation, at least on the first implementation iteration? Let me know your thoughts on it, please.

This would restrict the uses cases drastically. We would only work for multiple joins that use exactly the same key across all levels. The common partition key is already a somewhat hard restriction and if we allow multiple joins that join on the exact same keys, we'll probably address a little amount of real-world use cases. One more example that wouldn't also be supported `T1 JOIN T2 ON T1.key = T2.key JOIN T3 ON T1.key = T3.key  AND T2.k99 == T3.k99`. The multi-join will be disabled by default first, and users should enable it if their common joining key has a considerable distribution that matches their expected parallelism. That should be often the case for joins using primary keys, which have high cardinality. Another possibility in a second interaction would be creating more config options to control this, if we think it's necessary. What do you think?

gustavodemorais replied:
Hey Steve, you're right that we risk running into a OOM issue. We have a Map value for both InputSideHasUniqueKey and InputSideHasNoUniqueKey and could face this if we have too many rows for an specific join key. I'd like us to try to come up with a good solution for this so we support more use cases and ultimately fix the intermediate state issue for more users. I've been thinking about a solution to address this by creating a composite key and having only one level as I mentioned previously. 

For InputSideHasUniqueKey, for example, that means using MapState[(JoinKey, UniqueKey)]. We could then iterate through records one by one and then look at the key. We'd only have one key and value at a time in memory. We'd have to go through all common keys but we would [only deserialize the key](https://github.com/apache/flink/blob/f2568dee63138899cb80982a9659ab25f0d38c2c/flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/state/rocksdb/RocksDBMapState.java#L499) and go to the next key if that doesn't match our join condition. We only deserialize the value if the key matches our joinKey. That would solve the memory issue and would only lightly impact performance since we filter on keys directly.

For InputSideHasUniqueKey we'd have MapState[(JoinKey, Row)], which means we would also deserialize the value  for each entry. That means a more meaningful impact but we'd also only iterate one at a time and the memory issue would also be solved

This strategy would work the same as having commonKey = joinKey and we'd also supporting more use cases - if can users enabled multi joins. Let me know what you think @SteveStevenpoor 🙂

SteveStevenpoor replied:
Unfortunately you can't use MapState[joinKey, UniqueKey]. For example, you have pairs of (JoinKey, UniqueKey) and put (1,1), then (1,2) into rocksdb. Only (1,2) will be considered alive to rocksdb and (1,1) will be overwritten. 

Same for MapState[JoinKey, Row]. There will be only one row for each join key which leads to data loss. Only last update of key's value considered alive.

I don't currently see an easy way to implement this correctly without reintroducing some form of nested mapping or custom structure to handle multiple values per key. I do think it'd be ok if we stick to simple cases (commonKey=joinKey) at least at first.

gustavodemorais replied:
> Unfortunately you can't use MapState[joinKey, UniqueKey]. For example, you have pairs of (JoinKey, UniqueKey) and put (1,1), then (1,2) into rocksdb. Only (1,2) will be considered alive to rocksdb and (1,1) will be overwritten

How would they override the value if we have MapStateDescriptor<[joinKey, UniqueKey], RowData> recordStateDesc? They are ultimately different keys and after adding both to state we'd have {[1,1] -> val1, [1,2] -> val2}.

gustavodemorais replied:
Hey Stepan, thank you also for the quick responses. 

I've implemented my suggestion. Take a look https://github.com/apache/flink/pull/26313/commits/685af0c6d7edc1e0c89b6c84ac3185dc67a05d05. That should fix the memory issue and works mostly the same for commonKey = joinKey with only one level 🙂

gustavodemorais replied:
> Given that, maybe we can take two parallel tracks?

As I already mentioned, we have some stuff left to make a first end to end version work and it'd be more productive to have us working on different parts instead of on the same code while supporting each other with reviews. With a working e2e version, we can work on some benchmarks and make more informed changes based on numbers. If you have something for me to look at for the optimizer, let me know.

SteveStevenpoor replied:
> Our Multi-Join Operator has the common key as the operator key, and not necessarily the join key, so we need to have MapState<JoinKey, Map<Row, Count>>.

So try to test your implementation without JoinKey = CommonKey. For example, In StreamingMultiJoinOperatorTestBase change keySelector from joinKeySelector to something like: 
<pre>
            harness.setKeySelector(i, (KeySelector<RowData, RowData>) data -> {
                GenericRowData key = new GenericRowData(1);
                key.setField(0, data.getString(1));
                System.out.println(key);
                return key;
            });</pre>
            
I tried it and tests failed, but maybe it need a little bit more work with the test. Moreover, please take a closer look on the example I provided above. In general, if we partition on commonKey != JoinKey we can't provide consistency since in MapState we iterate only inside partition key. So if joinKeys of two records are equal and commonKeys are not, you will get incorrect results.

> As I already mentioned, we have some stuff left to make a first end to end version work and it'd be more productive to have us working on different parts instead of on the same code while supporting each other with reviews. With a working e2e version, we can work on some benchmarks and make more informed changes based on numbers. If you have something for me to look at for the optimizer, let me know.

I agree in principle, but for me to work on the optimizer side, I need clarity on the partitioning strategy. At this point, I don’t see a viable path forward using commonKey != joinKey, for all the reasons outlined above. I’m happy to contribute, but I’d prefer to build on a strategy where we have consistency guarantees from the start — for instance, constraining the first version to commonKey == joinKey.

Let me know what you think.

gustavodemorais replied:
One positive thing I've looked into about dropping one join condition and using a commonJoinKey (JoinKey != CommonKey): when we use a subset of the join keys, we locally have all associated data with a join for all N input streams. That means, we avoid network shuffle between operators instances and taskmanagers which is also responsible for a big overhead. Just something I was looking into and wanted to share here.

rovboyko replied:
Hi @gustavodemorais @SteveStevenpoor !
Maybe my comment is too late but I think I need to say a couple of words about using CommonJoinKey as a subset of JoinKey.
I agree with Gustavo - using the CommonJoinKey helps us to get all needed records on correct subtask thus the data consistency in not broken.

But in other hand we'll have the following example `SELECT * FROM T1 JOIN T2 on T1.k1 = T2.k1 AND **T1.k98 = T2.k98** JOIN T3 on T2.k1 = T3.k1 AND T2.k99 == T3.k99`.
And if we receive the input data from T1 (with particular values of T1.k1 and T1.k98) we need first to get all data from T2 and then request from T3 the whole set of T2.k1 and T2.k99 which we found on previous step. So there would be as many requests to T3 as many different values of k99 were found in T2. And the performance would decrease **exponentially** depending on input tables number.
And more over, the future optimization which might be used for MJ is changing state order in runtime would not be possible to use if the JoinKey != CommonKey. Or we can use parallel requests to different states in future. But using CommonJoinKey will not allow such optimizations.

So I'm tending to agree with Stepan's idea in comments above. Where he offered to not use the CommonJoinKey instead of JoinKey. I think such approach will slightly reduce the use cases but will bring much better performance (we won't need an exponential traverse across N states, just one iteration through each state) and, which is more important, the ability for future optimizations.

rovboyko replied:
Hi @gustavodemorais ! Thank you so much for your reply and explanation!

> It's not necessary to get all the data from T2.
Yes, I didn't mean we need to get **all** data from T2, but only all associated with K1 records.

> For example, for SELECT * FROM T1 JOIN T2 on T1.k1 = T2.k1 AND T1.k98 = T2.k98 JOIN T3 on T2.k1 = T3.k1 AND T2.k99 == T3.k99 we could start at any position.

In this example: If are receiving input record from T1 and starting from T3 then you can't filter the records on T2.k99 == T3.k99 condition because you don't have the k99 inside T1. So the states requesting order changing won't work. For the same reason you won't have an ability to request multiple states in parallel.

>  Let me give a classic example SELECT * FROM users LEFT JOIN orders on users.id = orders.uid LEFT JOIN payment orders.uid = payment.uid AND orders.id = payment.oid -> this would already not be supported and it's only a 3 way join!
Yes, I agree with you. But this is only the question of DB normalization. In normalized DB you won't have field `payment.uid` in table `payment`. And of course if you will join 5,10,15 tables you most probably will not have a even a CommonJoinKey among all tables. And from my point of view - the only DB schema where the Multi Way Join can be implemented is `star` schema. I don't try to say that using only one JoinKey as a multiWayJoin condition is enough for many cases. But I think that using commonJoinKey instead is not much wider.

> All that said, when we have a first e2e working version, we can check real benchmark numbers to see the impact of this and if there's a significant optimization we won't be able to implement. If that's the case, I think the community would be happy to receive one more operator contribution/optimizations. The optimizer could identify the specific cases and then decide accordingly which operator to use. Let me know if that makes sense :)
Yes, thank you for this clarification. Let's wait for benchmark results. And if it can be improved by implementing additional version of StreamingMultiJoinOperator (which would work only for JoinKey, not CommonJoinKey) then we will try to do it. 

BTW here you can find the PR to nexmark project with attempt of constructing and benchmarking multi way join - https://github.com/nexmark/nexmark/pull/70

dawidwys replied:
Since all the cases we see in our production use cases have additional joining key, I'd suggest we start with what Gustavo implemented. We can prepare a special optimised scenario for when commonkey = joinkey later.

gustavodemorais replied:
Thanks for the reply and the valuable discussion, @rovboyko and @SteveStevenpoor! Glad to have your eyes on the matter. If you want to contribute to any one of the following tickets I'll be opening, let me know. I'll try to finish the feature by June 24th. 

Otherwise, glad to get hear your thoughts in other reviews.

snuyanzin commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,830 @@
+        // if the corresponding left-side row previously had no matches (indicated by `!matched`),
+        // it would have resulted in a null-padded output. This new record might form a
+        // valid join, so the previous null-padded output must be retracted.
+        if (isUpsert(input) && isLeftJoin && !matched) {
```
nit: I wonder if we can reuse `UpdateMode.UPSERT`, `RETRACT` here to simplify the code a bit...

gustavodemorais replied:
I searched the codebase for updateMode.UPSERT and found this in Plannerbase
```
val (needUpdateBefore, withChangeFlag) = outputConversion.getUpdateMode match {
          case UpdateMode.RETRACT => (true, true)
          case UpdateMode.APPEND => (false, false)
          case UpdateMode.UPSERT => (false, true)
        }
```

Can you give me a simple example on how we could use this here?

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,830 @@
+     * See {@link StreamingMultiJoinOperator} for a detailed explanation of the recursive join and
+     * examples.
+     *
+     * @param depth The current depth of the recursion, representing the input stream index (0 to
```
I've touched on this on the FLIP here: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=350784112#FLIP516:MultiWayJoinOperator-RecursivevsIterative. Take a look and let me know what you think

SteveStevenpoor replied:
Thanks for pointing me the section. I agree with you that a depth search first across join inputs does align with a recursive structure.

That said, I’ve actually implemented a similar multi-join operator with an iterative approach in our internal fork, and it worked fine in performance and the logic remained quite readable. Also there is a possibility that in the future we will want to change traversal order. In my mind it would be more simple with maps.

I totally understand that this a matter of preference, but I just wanted to share my experience in case it can help. If you're interested in revisiting this direction, I can help you.

gustavodemorais replied:
Hey Steve, thanks for your input. I totally understand where you are coming from. I think my personal preference would be sticking to the recursive approach for now since I don't see any technical limitations + think the code will look cleaner, even though I do believe both are good approaches.

We still have a lot of missing tasks to get the multi join to a running state. I think right now, my idea would now be to address the missing pieces. We can always reevaluate this in a second interaction or as part of another ticket. As long as the layout of the operator doesn't change, it shouldn't be a problem changing the implementation later, if necessary.

gustavodemorais commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperatorTestBase.java:
```diff
@@ -0,0 +1,644 @@
+                        .toArray(new LogicalType[0]));
+    }
+
+    protected static GeneratedMultiJoinCondition createMultiJoinCondition(int numInputs) {
```
That was a really good pointer, thanks @dawidwys. I refactored all conditions to use this pattern and like it much better.

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,830 @@
+            throws Exception {
+
+        boolean isLeftJoin = isLeftJoinAtDepth(depth);
+        RowKind inputRowKind = input.getRowKind(); // Preserve original RowKind
```
Yep, removed!

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,830 @@
+        // This optimization avoids that by skipping state iteration at the input's own level.
+        // For inner joins, the input record itself will be processed at the `inputId`
+        // level in `processInputRecord`. We do not have to process any records in state for
+        // the inputId if it's an inner join, since we do not need to know if we the left
+        // side has associations. We only need to know if the left side has associations if
```
I've rewritten it. Take a look if makes it better, please. Let me know if not

gustavodemorais commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperator.java:
```diff
@@ -0,0 +1,830 @@
+    }
+
+    // Problem: we are not using the primary key but rather always the partitioning key
+    private boolean processRecords(
```
Hmmm, I get where you're coming, but I don't think so. The main goal of this function is to iterate through records in state for each depth using recursion with a depth first search approach. Then, it also returns if it actually matched a combination because we need that info. Let me know if you have other naming suggestions

gustavodemorais commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/operators/join/stream/StreamingMultiJoinOperatorTest.java:
```diff
@@ -0,0 +1,1257 @@
+import java.util.List;
+import java.util.Map;
+
+class StreamingTwoWayInnerMultiJoinOperatorTest extends StreamingMultiJoinOperatorTestBase {
```
All of these were added. Also a four way mixed join that has inner joins, left joins and multiple conditions for one join with equi and non-equi.

rkhachatryan commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/MultiJoinStateViews.java:
```diff
@@ -0,0 +1,440 @@
+                            return true;
+                        }
+                    }
+                    return false;
```
Can we terminate the iteration as soon as the record doesn't match join condition? I don't think so.
IIUC, `recordState` might look like this:
```
// JK, UK, value
 <0,   a,    ...>
 <1,   b,     ...>
 <1,   c,     ...>
 <2,   d,     ...>

(this is already with some common key selected)

So for JoinKey 1, we need to collect two records in the middle.
(also, there're no guarantees that the records are sorted in general)

Or am I missing something?

rkhachatryan replied:
I'm pretty sure that this optimization would be required, but I still might be wrong :)
It would be great to add a benchmark to https://github.com/apache/flink-benchmarks and compare the performance of the new approach with the current one.
This might also be helpful to judge about https://github.com/apache/flink/pull/26313#discussion_r2105804426

gustavodemorais replied:
> I'm pretty sure that this optimization would be required, but I still might be wrong :)

How could we implement this RocksDB-optimized version - is this an operator decision? If you have any pointers to where we do something similar, would be happy to look into those.

> It would be great to add a benchmark to https://github.com/apache/flink-benchmarks

Yes, there are multiple smaller optimizations that I have in my todo list to try out but wanted to do them only after I'm able to measure things. Basically following one of the "general remarks" from the benchmarks repo "If you can not measure the performance difference, then just don't bother (avoid premature optimisations)"

gustavodemorais replied:
Ok, I've looked into this. In short, we'll need a SortedMapState to implement this optimization. With that, we'd be able to seek the first matching keys and also exit early when we have no more matching join keys. This is definitely a considerable optimization.

We'd need something like this https://cwiki.apache.org/confluence/display/FLINK/FLIP-220%3A+Binary+Sorted+State and could revisit as soon as we have it.

rkhachatryan commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/MultiJoinStateViews.java:
```diff
@@ -0,0 +1,440 @@
+        public void addRecord(RowData joinKey, RowData record) throws Exception {
+            RowData uniqueKey = uniqueKeySelector.getKey(record);
+            RowData compositeKey = createCompositeKey(joinKey, uniqueKey);
+            recordState.put(compositeKey, record);
```
Should we project the input record and only store the needed fields to reduce space consumption and IO?

gustavodemorais replied:
That's an interesting idea. Currently, a StreamingJoinOperator is locally chained to a Calc inside of a Task which is kind of a similar optimization. Your suggestion is basically that we merge both operators and that the MJOperator takes care of materializing (push down the optimization), if I understand it correctly. 

In theory it makes sense. From an optimizer perspective we'd definitely need a couple of changes. I wonder why the other operators do not do this - maybe there's a drawback/reason behind it? I'd either way suggest addressing this as part of another ticket since it's a separate optimization. But I think it's something we should look into.

rkhachatryan replied:
I think it's all about priorities, I know a couple of other places where projections could be pushed down as well, but  so far was not prioritized or was addressed differently.

However, in this case it **_might_** be more pressing because the amount of state to process a single record might be bigger.

Addressing this in a separate ticket (if we decide to do so) would make sense, but we need to keep in mind that the state would have to be migrated (or we support this only for the new jobs somehow).

rkhachatryan replied:
> However, in this case it might be more pressing because the amount of state to process a single record might be bigger.

There will be N states for N inputs, and the number of reads might grow exponentially (depending on the "branching" factor, i.e. how many matches do we have for a single record).
That might exhaust RocksDB block cache (especially given that it doesn't separate keys and values).
So at a first glance, this optimization also seems to be highly desirable (but again, good to have some benchmark results).

rkhachatryan commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/MultiJoinStateViews.java:
```diff
@@ -0,0 +1,440 @@
+                        Map.Entry<RowData, Integer> currentEntry = stateIterator.next();
+                        RowData compositeKey = currentEntry.getKey();
+                        RowData currentJoinKey = compositeKey.getRow(0, joinKeyType.getArity());
+                        if (joinKey.equals(currentJoinKey)) {
```
Here, we assume that `equals` yields the same as equality predicate in JOIN. 
Similarly, we assume that for binary forms of keys (when we use them with RocksDB in map state).
Does this assumption always hold? 


cc: @dawidwys

rkhachatryan replied:
Running the tests with RocksDB should be easy; and I'd say it's mandatory to run the tests with the both state backends (the usual way to do that is to parameterize them).

I can look deeper into it later; I would also be great to hear Dawids opinion

gustavodemorais replied:
We've synced and I'm adding parameterized tests for rocksdb and changing how we check the joinKey equality to make it general 👍

rkhachatryan replied:
Awesome, thanks!

rkhachatryan reviewed (commented):
Thanks for the PR Gustavo!
I've asked some questions on the PR 
(I lack some context, so not all might not be relevant).

gustavodemorais commented:
> Thanks for the PR Gustavo! I've asked some questions on the PR (I lack some context, so not all might not be relevant).

Thanks for looking into the state design, @rkhachatryan. Replied to the comments. If you have any feedback on the operator itself or any other part of the code, happy to hear them

dawidwys reviewed (approved):
The PR lgtm. The comments are really helpful now!

I'll give a chance to @snuyanzin and @rkhachatryan to leave their last comments and I'll merge on monday.
